### ISSUE TITLE
Multiple outputs for regression

### ISSUE THREAD

**Issue by IvanPetrovMck** [microsoft/LightGBM#3500]
Hi,

I'm trying to predict yields that sum to 100 percents... Is there a possibility to do this in lightgbm? Basically it is a multioutput regression that have softmax layer and mse loss. I've tried using multiclass classifier as a base and wrote a custom objective. But apparently this doesn't work... I've checked derivatives of objective and they are correct. So most probably the reason is in trees which give single prediciton instead of probs for each leaf. Is there any other way to do this? Thanks 

```
def my_logistic_obj(y_hat, dtrain): 
    
    y_true = y.values
    n_data = y_true.shape[0] 
    n_classes = y_true.shape[1]
    
    ## get yields
    y_hat = np.reshape(y_hat, [n_data, n_classes])
    y_hat = np.exp(y_hat)
    
    
    probs = y_hat / np.sum(y_hat,axis=1).reshape([-1,1])

    L = (y_true - probs)**2
    L = L.sum(axis=1)
    
    grad = np.zeros_like(probs) * 1.0
    hess = np.zeros_like(probs) * 1.0

    for k in range(n_classes):
        I = np.zeros ([n_data,n_classes])
        I[:,k] = 1
        prob_k = probs[:,k].reshape([-1,1])
        prob_k_hat = y_true[:,k].reshape([-1,1])        
        ss = 2 * (prob_k - prob_k_hat) * prob_k * (I - probs )
        grad += ss

        ss = 2 * (2 * prob_k - prob_k_hat) * prob_k * np.power(I - probs,2) + \
             2 * (prob_k * prob_k - prob_k * prob_k_hat) * (-probs) * (1-probs)
        hess += ss
    grad = grad / n_data
    hess = hess / n_data

    return grad.flatten("F"), hess.flatten("F")

categoricals = None
lgb_train = lgb.Dataset(X) 

params = {
        'seed': 15,
        'boosting_type': 'gbdt', ## rf, gbdt, dart
        'num_class': y.shape[1],
        'max_depth': 25,
        'learning_rate': 0.05,
        'feature_fraction': 1,
        'bagging_fraction': 0.8,
        'bagging_freq': 1,
        'verbose': 1,
        'rate_drop': 0.1,
        'skip_drop': 0.5,
        'num_leaves': 10,
        'lambda_l2': 0
                }
    
evals_result = {}
gbm = lgb.train(params,
                lgb_train,
                num_boost_round = 200,
                valid_sets=[lgb_train],
                fobj = my_logistic_obj,
                verbose_eval=1
               )
```

**Comment by guolinke** [comment]
maybe you can use https://scikit-learn.org/stable/modules/generated/sklearn.multioutput.MultiOutputRegressor.html

**Comment by IvanPetrovMck** [comment]
guolinke this wouldn't work, because I want to have all coefficients summed up to 100%

**Comment by guolinke** [comment]
@IvanPetrovMck 
Use multi-class as base, and customed objective function should work.

However, the probability is not given by the leaves in single tree, it uses one tree for one class. 
For example, assuming there are 3 classes, A, B and C. And trees are orderder like A1, B1, C1, A2, B2, C2, ...
And the score for class A is sum(A1, A2, ...), and the same for B and C.

BTW, you should provide the gradient and hessian for each class too. And the order of gradient and hessian is different with trees. For example, `grad[i * k + j]` means the gradient for `j`-th class of `i`-th sample, and the total class number is `k`.

**Comment by IvanPetrovMck** [comment]
@guolinke Does same state for 'preds'? Are they also ordered in [n_classes, n_datapoints]?

**Comment by IvanPetrovMck** [comment]
Ок, everything works. The problem was with dimensions of gradients and predictions. Here is full working code for history. Thanks for helping.

```
def my_softmax_rmse(y_hat, dtrain): 
    ##
    ##
    """
        Softmax probabilities with RMSE loss
        :y_hat: y_hat is predictions, if 1D list that reflects [n_classes, n_observations]
         
    """
    y_true = y.values.T
    n_classes = y_true.shape[0]    
    n_data = y_true.shape[1] 
      

    ## get yields
    y_hat = np.reshape(y_hat, [n_classes, n_data])
    
    
    y_hat = y_hat + np.log(y_true.mean(axis=1)).reshape([-1,1])
    y_hat = np.exp(y_hat)
    
    probs = y_hat / np.sum(y_hat,axis=0).reshape([1,-1])
    
    L = (y_true - probs)**2
    L = L.sum()
    print (L)
    
    grad = np.zeros_like(probs) * 1.0
    hess = np.zeros_like(probs) * 1.0

    for k in range(n_classes):
        
        I = np.zeros ([n_classes, n_data])
        I[k,:] = 1
        
        prob_k = probs[k,:].reshape([1,-1])
        prob_k_hat = y_true[k,:].reshape([1,-1])   
        
        ss = 2 * (prob_k - prob_k_hat) * prob_k * (I - probs )
        grad += ss

        ss = 2 * (2 * prob_k - prob_k_hat) * prob_k * np.power(I - probs, 2) + \
             2 * (prob_k * prob_k - prob_k * prob_k_hat) * (-probs) * (1-probs)
        hess += ss
        
    grad = grad 
    hess = hess 
    
    return grad.flatten(), hess.flatten()
```

**Comment by HHHabcd** [comment]
I'm working on a same problem recently. I think your code is quite effective. But I'm confused what the 'ss' means, and what the fomula:
"ss = 2 * (prob_k - prob_k_hat) * prob_k * (I - probs )" and 
"ss = 2 * (2 * prob_k - prob_k_hat) * prob_k * np.power(I - probs, 2) + 2 * (prob_k * prob_k - prob_k * prob_k_hat) * (-probs) * (1-probs)"
is calculating?
